File: Typecast/CFFCharset.m
//
//  CFFCharset.m
//  Type Designer
//
//  Created by David Schweinsberg on 4/01/13.
//  Copyright (c) 2013 David Schweinsberg. All rights reserved.
//

#import "CFFCharset.h"
#import "TCDataInput.h"

@implementation CFFCharset

- (int)format
{
    return -1;
}

- (int)SIDForGID:(int)gid
{
    return 0;
}

@end


@interface CFFCharsetFormat0 ()
{
    NSMutableArray *_glyphs;
}

@end

@implementation CFFCharsetFormat0

- (id)initWithDataInput:(TCDataInput *)dataInput glyphCount:(int)glyphCount
{
    self = [super init];
    if (self)
    {
        _glyphs = [[NSMutableArray alloc] initWithCapacity:glyphCount - 1]; // minus 1 because .notdef is omitted
        for (int i = 0; i < glyphCount - 1; ++i)
            [_glyphs addObject:[NSNumber numberWithUnsignedShort:[dataInput readUnsignedShort]]];
    }
    return self;
}

- (int)format
{
    return 0;
}

- (int)SIDForGID:(int)gid
{
    if (gid == 0)
        return 0;
    return [_glyphs[gid - 1] intValue];
}

@end


@interface CFFCharsetFormat1 ()
{
    NSMutableArray *_charsetRanges;
}

@end

@implementation CFFCharsetFormat1

- (id)initWithDataInput:(TCDataInput *)dataInput glyphCount:(int)glyphCount
{
    self = [super init];
    if (self)
    {
        _charsetRanges = [[NSMutableArray alloc] init];

        int glyphsCovered = glyphCount - 1;  // minus 1 because .notdef is omitted
        while (glyphsCovered > 0)
        {
            int first = [dataInput readUnsignedShort];
            int left = [dataInput readUnsignedByte];
            NSRange range = NSMakeRange(first, left);
            [_charsetRanges addObject:[NSValue valueWithRange:range]];
            glyphsCovered -= left + 1;
        }
    }
    return self;
}

- (int)format
{
    return 1;
}

- (int)SIDForGID:(int)gid
{
    if (gid == 0)
        return 0;

    // Count through the ranges to find the one of interest
    int count = 0;
    for (NSValue *rangeValue in _charsetRanges)
    {
        NSUInteger first = [rangeValue rangeValue].location;
        NSUInteger left = [rangeValue rangeValue].length;
        if (gid < left + count)
        {
            NSUInteger sid = gid - count + first - 1;
            return (int)sid;
        }
        count += left;
    }
    return 0;
}

@end


@interface CFFCharsetFormat2 ()
{
    NSMutableArray *_charsetRanges;
}

@end

@implementation CFFCharsetFormat2

- (id)initWithDataInput:(TCDataInput *)dataInput glyphCount:(int)glyphCount
{
    self = [super init];
    if (self)
    {
        _charsetRanges = [[NSMutableArray alloc] init];

        int glyphsCovered = glyphCount - 1;  // minus 1 because .notdef is omitted
        while (glyphsCovered > 0)
        {
            int first = [dataInput readUnsignedShort];
            int left = [dataInput readUnsignedShort];
            NSRange range = NSMakeRange(first, left);
            [_charsetRanges addObject:[NSValue valueWithRange:range]];
            glyphsCovered -= left + 1;
        }
    }
    return self;
}

- (int)format
{
    return 2;
}

- (int)SIDForGID:(int)gid
{
    if (gid == 0)
        return 0;

    // Count through the ranges to find the one of interest
    int count = 0;
    for (NSValue *rangeValue in _charsetRanges)
    {
        NSUInteger first = [rangeValue rangeValue].location;
        NSUInteger left = [rangeValue rangeValue].length;
        if (gid < left + count)
        {
            NSUInteger sid = gid - count + first - 1;
            return (int)sid;
        }
        count += left;
    }
    return 0;
}

@end

File: CFF/CFF/CFFCharset.swift
//
//  CFFCharset.swift
//  Type Designer
//
//  Created by David Schweinsberg on 8/10/17.
//  Copyright © 2017 David Schweinsberg. All rights reserved.
//

import Foundation
import IOUtils

protocol CFFCharset {
  var format: Int { get }
  func sid(gid: Int) -> Int
}

class CFFCharsetFormat0: CFFCharset {
  let glyphs: [Int]

  init(dataInput: TCDataInput, glyphCount: Int) {
    var glyphs = [Int]()
    for _ in 0..<glyphCount - 1 {  // minus 1 because .notdef is omitted
      glyphs.append(Int(dataInput.readUInt16()))
    }
    self.glyphs = glyphs
  }

  var format: Int {
    get {
      return 0
    }
  }

  func sid(gid: Int) -> Int {
    if gid == 0 {
      return 0
    }
    return glyphs[gid - 1]
  }

}

class CFFCharsetFormat1: CFFCharset {
  let charsetRanges: [CountableClosedRange<Int>]

  init(dataInput: TCDataInput, glyphCount: Int) {
    var charsetRanges = [CountableClosedRange<Int>]()

    var glyphsCovered = glyphCount - 1  // minus 1 because .notdef is omitted
    while glyphsCovered > 0 {
      let first = Int(dataInput.readUInt16())
      let left = Int(dataInput.readUInt8())
      let range = first...first + left
      charsetRanges.append(range)
      glyphsCovered -= left + 1
    }
    self.charsetRanges = charsetRanges
  }

  var format: Int {
    get {
      return 1
    }
  }

  func sid(gid: Int) -> Int {
    if gid == 0 {
      return 0
    }

    // Count through the ranges to find the one of interest
    var count = 0
    for range in charsetRanges {
      if gid <= range.count + count {
        let sid = gid - count + range.lowerBound - 1
        return sid
      }
      count += range.count
    }
    return 0
  }

}

class CFFCharsetFormat2: CFFCharset {
  let charsetRanges: [CountableClosedRange<Int>]

  init(dataInput: TCDataInput, glyphCount: Int) {
    var charsetRanges = [CountableClosedRange<Int>]()

    var glyphsCovered = glyphCount - 1  // minus 1 because .notdef is omitted
    while glyphsCovered > 0 {
      let first = Int(dataInput.readUInt16())
      let left = Int(dataInput.readUInt16())
      let range = first...first + left
      charsetRanges.append(range)
      glyphsCovered -= left + 1
    }
    self.charsetRanges = charsetRanges
  }

  var format: Int {
    get {
      return 2
    }
  }

  func sid(gid: Int) -> Int {
    if gid == 0 {
      return 0
    }

    // Count through the ranges to find the one of interest
    var count = 0
    for range in charsetRanges {
      if gid <= range.count + count {
        let sid = gid - count + range.lowerBound - 1
        return sid
      }
      count += range.count
    }
    return 0
  }

}
